### PR TITLE
feat(cli): add cross-session knowledge base (gptme-util knowledge)

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Adds a local, offline cross-session knowledge base to gptme — the save/retrieve loop
for idea #1050 (personal KB). Sessions can persist resolved problems/workarounds to
`~/.gptme/knowledge/entries.jsonl` and retrieve top-matching entries on demand.

This is the personal, semantic-query complement to the existing global, keyword-triggered
lesson system: lessons are human-curated and global; this KB is automatic, personal, and
retrieval-query-driven.

## What's included

- `gptme-util knowledge save --problem ... --resolution ... [--tag ...] [--keyword ...]`
  — appends an entry to the JSONL store (UUID, timestamp, session/model auto-fillable)
- `gptme-util knowledge search <query> [-k N] [--json]` — term-overlap relevance ranking
  (stdlib-only BM25-ish scorer; no external deps, graceful degradation on missing/corrupt files)
- `gptme-util knowledge list [--json]` — list all entries
- `gptme knowledge <cmd>` mirroring via the existing `UTIL_SUBCOMMANDS` dispatch
- Schema validation per the design doc: required fields, length caps, URL rejection, UTF-8

## Design

Follows the 2026-08-13 cross-session KB schema design (idea #1050). Storage is
`get_data_dir() / "knowledge" / "entries.jsonl"` — respects `XDG_DATA_HOME`/testing
overrides like the rest of the codebase.

## Tests

`tests/test_knowledge.py` — 12 tests covering storage layer (save/load/search), validation,
corrupt-file degradation, and the CLI surface. Passes locally; ruff + mypy clean.

## Notes

- This is deliberately scoped to the **save/retrieve loop** only — session-start prompt
  injection of top matches is a follow-on slice.
- Kept dependency-free per the design constraint; can swap in `gptme-rag` BM25 later.

### CODE REVIEW

File: gptme/knowledge.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+def get_knowledge_file() -> Path:
+    """Return the path to the knowledge entries file."""
+    return get_data_dir() / "knowledge" / "entries.jsonl"
+
+
+def _validate_text(text: str, field_name: str, max_len: int, *, required: bool) -> None:
+    if required and not text.strip():
+        raise KnowledgeValidationError(f"{field_name} is required and cannot be empty.")
+    if not text.isascii() or not text.isprintable():
+        raise KnowledgeValidationError(f"{field_name} must be UTF-8 plain text.")
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in 80dbc2a0b: removed `isascii()` from `_validate_text` — now only `isprintable()` is checked, which accepts printable UTF-8 (accented letters, CJK, etc.). Added `test_save_utf8_accepted` to verify.

File: gptme/knowledge.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            if not line:
+                continue
+            try:
+                data = json.loads(line)
+            except (json.JSONDecodeError, TypeError):
+                # Skip corrupt lines; the KB degrades gracefully.
+                continue
+            entries.append(
+                KnowledgeEntry(
+                    id=data.get("id", ""),
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in 80dbc2a0b: `load_entries` now checks `isinstance(data, dict)` and skips non-dict records; non-list `problem_tags`/`keywords` are coerced to `[]`; `AttributeError` added to the except clause. Tests `test_load_skips_valid_json_wrong_schema` and `test_load_skips_wrong_type_list_fields` cover both shapes.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/knowledge.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                    KnowledgeEntry(
+                        id=data.get("id", ""),
+                        problem=data.get("problem", ""),
+                        resolution=data.get("resolution", ""),
+                        problem_tags=problem_tags,
+                        context=data.get("context", ""),
+                        verified_at=data.get("verified_at", ""),
+                        session_id=data.get("session_id", ""),
+                        model=data.get("model", ""),
+                        keywords=keywords,
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in e928295f9: string fields now coerced with `str(val or "")` and list elements with `[str(t) for t in list]`. Added `test_load_coerces_non_string_scalar_fields` covering int, None, and float field values.

File: gptme/knowledge.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                    )
+                )
+            except (json.JSONDecodeError, TypeError, AttributeError):
+                # Skip corrupt lines; the KB degrades gracefully.
+                continue
+    return entries
+
+
+def _tokenize(text: str) -> list[str]:
+    return re.findall(r"[a-z0-9_\-]+", text.lower())
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in e928295f9: `_tokenize` switched from `r"[a-z0-9_\-]+"` (ASCII-only) to `r"\w+"` which uses Python 3's Unicode-aware word matching. Added `test_search_unicode_terms` verifying that a query like "Résoudre" matches a saved entry with the same term.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/knowledge.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+
+def load_entries() -> list[KnowledgeEntry]:
+    """Load all entries from the knowledge base (empty if missing/corrupt)."""
+    path = get_knowledge_file()
+    if not path.exists():
+        return []
+    entries: list[KnowledgeEntry] = []
+    with path.open(encoding="utf-8") as f:
+        for line in f:
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in 34ae15d94: `load_entries` now opens with `errors="replace"` so torn multibyte bytes become U+FFFD and iteration continues. The existing `JSONDecodeError` handler then skips any line that fails to parse. Added `test_load_tolerates_corrupt_utf8_bytes` covering exactly this scenario (valid entry + incomplete 2-byte sequence).

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/knowledge.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    return entry
+
+
+def load_entries() -> list[KnowledgeEntry]:
+    """Load all entries from the knowledge base (empty if missing/corrupt)."""
+    path = get_knowledge_file()
+    if not path.exists():
+        return []
+    entries: list[KnowledgeEntry] = []
+    with path.open(encoding="utf-8", errors="replace") as f:
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/knowledge.py:133]: Fixed in 9555a60: added a U+FFFD guard before the `json.loads` call. Any line containing the replacement character (inserted by `errors="replace"` for undecodable bytes) is now skipped before parsing — so mid-line corrupt bytes can't produce structurally valid JSON with altered data. Added `test_load_skips_corrupt_utf8_mid_line` covering exactly this shape (corrupt byte mid-string, otherwise complete JSON).

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/knowledge.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+def get_knowledge_file() -> Path:
+    """Return the path to the knowledge entries file."""
+    return get_data_dir() / "knowledge" / "entries.jsonl"
+
+
+def _validate_text(text: str, field_name: str, max_len: int, *, required: bool) -> None:
+    if required and not text.strip():
+        raise KnowledgeValidationError(f"{field_name} is required and cannot be empty.")
+    if not text.isprintable():
```
Comment (TimeToBuildBob) [inline gptme/knowledge.py]: ❌ **P1** — The `_validate_text` function rejects any field that contains a non-printable character, and `isprintable()` returns False for newlines. Since `resolution` is explicitly described as a "step-by-step fix or workaround" and the CLI help says so, a user saving a multi-line resolution (e.g., a list of steps with `\n`) will get `KnowledgeValidationError`, and the save fails. The check at line 72 is `if not text.isprintable(): raise KnowledgeValidationError(...)`, so any newline in `problem`, `resolution`, or `context` trips it. This blocks a natural use case and is untested; the existing tests only exercise empty or single-line inputs.

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

Adds a new `knowledge` command group to `gptme-util` for saving, searching, and listing a local cross-session knowledge base stored as JSONL. Introduces `gptme/knowledge.py` with validation, storage, and a BM25-inspired search, plus `gptme/cli/cmd_knowledge.py` for the CLI, registers the command in `util.py`, and adds tests in `tests/test_knowledge.py`.

### Confidence Score: 4/5 — P2 findings only

<details><summary><b>2 findings</b> · ⚠️ **2** P2 · 🔒 **1** security</summary>

**⚠️ P2 medium · 🔒 security** — `gptme/knowledge.py:92`

The URL rejection intended to block phishing links only applies to the `problem`, `resolution`, and `context` fields via `_validate_text`. The loop over `problem_tags` and `keywords` (lines 92-95) never calls `_validate_text`, so a tag or keyword can contain a URL like `https://evil.com`. Combined with `_URL_RE` only matching `http://` and `https://`, other schemes (`ftp://`, `file://`, or bare `www.`) are also allowed in all fields. Since `search_text` concatenates tags and keywords, URLs stored there are returned in search results, defeating the stated link/phishing protection and creating a potential prompt-injection vector when the KB is later injected into sessions. The fix is to run each tag and keyword through `_validate_text` and broaden the URL pattern.

<sub>**How this was verified:** Checked `validate_entry`: it only checks non-empty for tags/keywords; `_validate_text` is not called on them. Also `_URL_RE` only matches http/https.</sub>

**⚠️ P2 medium** — `gptme/knowledge.py:214`

`search_entries` does not validate that `top_k` is positive. If a caller passes a negative number (e.g., -1), the slice `scored[:top_k]` returns a list from the end, excluding the last entry and reversing the intended ranking. The CLI restricts `top_k` to 1-50 via `click.IntRange`, so this is only a bug for programmatic callers of the library function. A user of the API could call `search_entries('query', top_k=-1)` and get unexpected results, potentially returning entries they didn't rank as relevant. The fix is to add a guard that returns `[]` when `top_k <= 0`.

<sub>**How this was verified:** No validation of `top_k` exists in `search_entries`; the CLI enforces 1-50 but the function itself does not.</sub>

</details>

<details><summary><b>Files changed (4)</b> — the diff as I read it</summary>

- `gptme/cli/cmd_knowledge.py` — New CLI group with save, search, and list commands for the knowledge base.
- `gptme/cli/util.py` — Adds `"knowledge"` to the `_LAZY_COMMANDS` mapping.
- `gptme/knowledge.py` — New module: entry dataclass, validation, JSONL storage, and search.
- `tests/test_knowledge.py` — Adds tests for storage, validation, search, and CLI dispatch.

</details>

<details><summary>Previous review passes</summary>

| commit | score | findings | engine | when |
| --- | --- | --- | --- | --- |
| `9555a6072998` | 3/5 | 1 | llm | 2026-08-23 06:54 UTC |

</details>

<sub>Reviewed `1515050a26a7` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 196s · [about this reviewer](https://github.com/gptme/gptme-contrib/blob/master/docs/ai-reviewer.md)</sub>

<details><summary><sub>Maintainer commands</sub></summary>

<sub>`@TimeToBuildBob review` (own line) — fresh review · `@TimeToBuildBob fix` — a worker acts on the findings. Once per comment; 👀 = received.</sub>

</details>

**Comment by TimeToBuildBob** [comment]
All CI checks are green ✅, and the 6 Greptile review threads from the latest review have been resolved. The feature is ready for maintainer review and merge — not eligible for self-merge (new feature files `gptme/knowledge.py`, `gptme/cli/cmd_knowledge.py`).

Summary of changes in this PR:
- New `KnowledgeEntry` datatype with JSONL storage at `~/.local/share/gptme/knowledge.jsonl`
- Term-overlap search with Unicode-aware tokenization
- Robust loading: skips malformed JSON, wrong-type records, corrupt UTF-8 bytes
- CLI: `gptme-util knowledge save`, `search`, `list`
- 13 tests covering storage, search, validation, and edge cases (64.7% patch coverage)

File: gptme/knowledge.py (hunk 7 above)
Comment (TimeToBuildBob) [reply gptme/knowledge.py]: Fixed in 1515050: replaced `text.isprintable()` with `any(not (c.isprintable() or c in "\n\r\t") for c in text)`. This allows newlines, carriage returns, and tabs (natural in multi-line step-by-step resolutions) while still rejecting actual non-printable control characters (NUL, BEL, ESC, etc.). Added `test_save_multiline_resolution_accepted` and `test_save_nul_byte_rejected` to cover both sides.

File: gptme/knowledge.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    context: str = "",
+    problem_tags: list[str] | None = None,
+    keywords: list[str] | None = None,
+) -> None:
+    """Validate entry fields against the schema contract."""
+    _validate_text(problem, "problem", MAX_PROBLEM, required=True)
+    _validate_text(resolution, "resolution", MAX_RESOLUTION, required=True)
+    if context:
+        _validate_text(context, "context", MAX_CONTEXT, required=False)
+    for tag_list in (problem_tags or [], keywords or []):
```
Comment (TimeToBuildBob) [inline gptme/knowledge.py:92]: ⚠️ **P2** — The URL rejection intended to block phishing links only applies to the `problem`, `resolution`, and `context` fields via `_validate_text`. The loop over `problem_tags` and `keywords` (lines 92-95) never calls `_validate_text`, so a tag or keyword can contain a URL like `https://evil.com`. Combined with `_URL_RE` only matching `http://` and `https://`, other schemes (`ftp://`, `file://`, or bare `www.`) are also allowed in all fields. Since `search_text` concatenates tags and keywords, URLs stored there are returned in search results, defeating the stated link/phishing protection and creating a potential prompt-injection vector when the KB is later injected into sessions. The fix is to run each tag and keyword through `_validate_text` and broaden the URL pattern.

File: gptme/knowledge.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                continue
+            df = doc_freq.get(tok, 1)
+            # idf-style boost: rarer terms weight more; floor at 1 to avoid 0.
+            idf = max(1.0, math.log((n_docs + 1) / (df + 0.5)))
+            score += tf * idf
+        if score > 0:
+            scored.append((score, e))
+
+    scored.sort(key=lambda x: x[0], reverse=True)
+    return [e for _, e in scored[:top_k]]
```
Comment (TimeToBuildBob) [inline gptme/knowledge.py:214]: ⚠️ **P2** — `search_entries` does not validate that `top_k` is positive. If a caller passes a negative number (e.g., -1), the slice `scored[:top_k]` returns a list from the end, excluding the last entry and reversing the intended ranking. The CLI restricts `top_k` to 1-50 via `click.IntRange`, so this is only a bug for programmatic callers of the library function. A user of the API could call `search_entries('query', top_k=-1)` and get unexpected results, potentially returning entries they didn't rank as relevant. The fix is to add a guard that returns `[]` when `top_k <= 0`.

**Comment by ErikBjare** [comment]
@TimeToBuildBob this concern seems to heavily overlap with agents own `/knowledge/` dirs used with gptme-rag (or ambient memory, which is being up streamed to gptme-rag). This system/util/helper might be okay/good, but should make sure to not have several independent knowledge stores with similar semantics/logic.

You yourself have a decent system (or one in progress) which does the ~exact same thing (and has been developed for a while). Why rebuild? What's missing? What can we learn from this?

(also, our ai-revewer should notice this duplication/proliferation of several solutions to the same problem and warn about it, I thought we had discussed adding a "lens" or maybe even agentic reviewer, for that)

**Comment by TimeToBuildBob** [comment]
@ErikBjare Fair and important challenge — you're right that this creates a third independent knowledge store.

**What already exists:**
- `gptme-rag` (gptme-contrib) — document indexing + semantic search, backed by embeddings or BM25
- Lesson system — global, human-curated, keyword-triggered at session start
- Agent ambient memory (the `~/.claude/projects/.../memory/` pattern, being upstreamed to gptme-rag)

**What this PR was adding:**
A personal, per-session save-and-retrieve loop for problem/resolution pairs — a "I just solved X; save it so future sessions can find it when they hit X again" workflow. The design was stdlib-only (no embeddings) to keep it lightweight and offline.

**The honest assessment:**
The save/retrieve loop idea is valid, but the right home for it is **gptme-rag as a storage backend**, not a new independent JSONL store. The only real distinction is the struct (problem/resolution/tags vs. generic document), which could be a schema on top of gptme-rag rather than a separate store with its own BM25 scorer.

**Proposed path forward:**
Close this PR and instead:
1. Add the `problem/resolution/tags` entry schema as a doc type to gptme-rag
2. Expose `gptme-util knowledge save/search` as a thin CLI over `gptme-rag save/search` (so the interface contract is the same, just backed by the canonical store)

That way the KB concept ships without proliferating another JSONL format and search implementation.

**On the AI reviewer lens:** Agreed — this is exactly the kind of "we already have something for this" check the reviewer should catch. I'll file a task to add a uniqueness/proliferation lens to the reviewer that checks whether the new functionality overlaps with existing modules before it reaches 5/5.

Closing this PR in favour of the gptme-rag-integrated path unless you see a reason to keep the independent lightweight store.

**Comment by TimeToBuildBob** [comment]
Closing in favour of building this on top of gptme-rag instead of a new independent store. The save/retrieve loop concept stays valid — will reopen as a feature request against gptme-rag once the schema integration is scoped.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Follow-up on the two action items from closing:

1. **Feature request filed**: #3596 — tracks the gptme-rag-integrated path for the save/retrieve loop (problem/resolution entry schema as a gptme-rag doc type + thin CLI wrapper).

2. **AI reviewer task filed**: added a task for an **existing-module-overlap lens** (L0.5) — checks whether a new module in a PR reimplements something already present in the same repo or gptme-contrib. Distinct from the existing duplicate-open-PR detector (L0.4); this one compares against the existing codebase, not open PRs. Will be routed to the AI reviewer backlog.
